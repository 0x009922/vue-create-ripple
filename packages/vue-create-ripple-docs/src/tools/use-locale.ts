import { ref, watchEffect, Ref } from "vue";

export interface UseLocaleOptions {
  available: string[];
  default: string;
}

export interface UseLocaleReturn {
  lang: Ref<string>;
}

export default function useLocale(opts: UseLocaleOptions): UseLocaleReturn {
  const KEY = 'locale';
  const initial = localStorage.getItem(KEY);

  // Язык браузера ('en-US' -> 'en')
  const browser = navigator.language.slice(0, 2);

  const lang = ref<string>(initial ?? browser);
  
  watchEffect(() => {
    // Сброс только к доступным, режим mandatory
    if (!opts.available.includes(lang.value)) {
      lang.value = opts.default;
    }

    // Сохранение в localStorage
    localStorage.setItem(KEY, lang.value);
  });


  return { lang };
}