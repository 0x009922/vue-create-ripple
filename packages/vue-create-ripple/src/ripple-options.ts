import { SetRequired } from 'type-fest';

export interface RippleOptions {
  /**
   * Effect disappearence timing function
   *
   * @default 'cubic-bezier(0.4, 0, 0.2, 1)'
   */
  disappearEasing?: string;

  /**
   * Effect disappearence duration
   *
   * @default '.7s'
   */
  disappearDuration?: string;

  /**
   * Effect appearence timing function
   *
   * @default 'cubic-bezier(0.16, 1, 0.3, 1)'
   */
  appearEasing?: string;

  /**
   * Effect appearence duration
   *
   * @default '.8s'
   */
  appearDuration?: string;

  /**
   * The class to be assigned to the circle. Can be used
   * for quick and easy styling.
   *
   * ! If not specified, then nothing will be visible
   */
  class?: string | null;

  /**
   * Flag to disable effect triggering
   */
  disabled?: boolean;

  /**
   * If `true`, the effect will always come from the center of the element.
   *
   * @default false
   */
  center?: boolean;

  /**
   * If `true`, then the effect will work only on event
   * `mousedown` on the element itself, not its descendants
   *
   * @default false
   */
  self?: boolean;

  /**
   * If `true`, then the effect when clicking on the child element will not work,
   * if in the tree of its parents there is an element bound to it
   * directive and this element is a child of the current one. Default
   * `true`, so you need to explicitly specify `false` to disable this option
   *
   * @default true
   */
  noNested?: boolean;

  /**
   * Additional css styles for the container
   */
  containerStyle?: Partial<CSSStyleDeclaration>;

  /**
   * Additional css styles for the ripple
   */
  circleStyle?: Partial<CSSStyleDeclaration>;

  /**
   * The absolute value of the maximum radius of the circle (in pixels).
   * By default, the circle is stretched enough to completely cover the entire
   * element.
   *
   * `null` means to stretch
   */
  radius?: number | null;
}

export type NormalizedRippleOptions = SetRequired<
  RippleOptions,
  | 'appearDuration'
  | 'appearEasing'
  | 'center'
  | 'disabled'
  | 'disappearDuration'
  | 'disappearEasing'
  | 'noNested'
  | 'radius'
  | 'self'
>;

const defaults: NormalizedRippleOptions = {
  disappearEasing: 'cubic-bezier(0.4, 0, 0.2, 1)',
  disappearDuration: '.7s',
  appearEasing: 'cubic-bezier(0.16, 1, 0.3, 1)',
  appearDuration: '.8s',
  class: null,
  disabled: false,
  center: false,
  self: false,
  noNested: true,
  radius: null
};

export function mergeDefaults(opts?: RippleOptions): NormalizedRippleOptions {
  return { ...defaults, ...opts };
}
