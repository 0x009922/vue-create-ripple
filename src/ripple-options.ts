export interface NormalizedRippleOptions {
  /**
   * Тайминг-функция исчезновения ripple.
   * По умолчанию 'cubic-bezier(0.4, 0, 0.2, 1)'
   */
  disappearEasing: string;
  /**
   * Длительность исчезновения ripple.
   * По умолчанию '.7s'
   */
  disappearDuration: string;
  /**
   * Тайминг-функция увеличения круга.
   * По умолчанию 'cubic-bezier(0.16, 1, 0.3, 1)'
   */
  appearEasing: string;
  /**
   * Длительность увеличения круга.
   * По умолчанию '.8s'
   */
  appearDuration: string;
  /**
   * Класс, который будет задан кругу. Можно использовать
   * для быстрой и простой стилизации.
   *
   * ! Если не будет указан, то ничего не будет видно
   */
  class?: string | null;
  /**
   * Если `true`, то эффект не будет срабатывать
   */
  disabled: boolean;
  /**
   * Если `true`, то эффект всегда будет исходить из центра элемента.
   */
  center: boolean;
  /**
   * Если `true`, то эффект сработает только при событии
   * `mousedown` на самом элементе, а не его потомках
   */
  self: boolean;
  /**
   * Если `true`, то эффект при клике на дочернем элементе не сработает,
   * если в дереве его родителей найдётся элемент с привязанной к нему
   * директивой и этот элемент является дочерним к текущему. По умолчанию
   * `true`, так что нужно явно указать `false` для отключения эффекта
   */
  noNested: boolean;
  /**
   * Объект для передачи стиля контейнеру напрямую.
   */
  containerStyle?: Partial<CSSStyleDeclaration>;
  /**
   * Объект для передачи стиля круга напрямую.
   */
  circleStyle?: Partial<CSSStyleDeclaration>;
  /**
   * Абсолютное значение максимального радиуса круга (в пикселях).
   * По умолчанию круг растягивается настолько, чтобы полностью покрыть весь
   * элемент.
   */
  radius: number | null;
}

export type RippleOptions = Partial<NormalizedRippleOptions>;

const defaults: NormalizedRippleOptions = {
  disappearEasing: 'cubic-bezier(0.4, 0, 0.2, 1)',
  disappearDuration: '.7s',
  appearEasing: 'cubic-bezier(0.16, 1, 0.3, 1)',
  appearDuration: '.8s',
  class: null,
  disabled: false,
  center: false,
  self: false,
  noNested: true,
  radius: null
};

export function mergeDefaults(opts?: RippleOptions): NormalizedRippleOptions {
  return { ...defaults, ...opts };
}
