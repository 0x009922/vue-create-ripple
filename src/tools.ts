/**
 * Расчёт конечного радиуса круга исходя из размеров
 * окна, где он будет, и месте клика мыши в окне
 *
 * @param width Ширина окна
 * @param height Высота окна
 * @param x X мыши
 * @param y Y мыши
 * @return Радиус
 */
export function computeRippleRadius(
  width: number,
  height: number,
  x: number,
  y: number
): number {
  const w = x > width / 2 ? x : width - x;
  const h = y > height / 2 ? y : height - y;
  return (w ** 2 + h ** 2) ** 0.5;
}

/**
 * Ожидание следующего фрема прорисовки
 * @param callback
 */
export function nextFrame(callback?: () => void): Promise<void> {
  return new Promise((resolve) => {
    requestAnimationFrame(() => {
      requestAnimationFrame(() => {
        resolve();

        if (callback) {
          callback();
        }
      });
    });
  });
}

/**
 * Расчёт координат мыши относительно элемента
 * @param el
 * @param e
 */
export function relativeMouseCoords(
  el: HTMLElement,
  e: {
    clientX: number;
    clientY: number;
  }
): [number, number] {
  const rect = el.getBoundingClientRect();
  return [e.clientX - rect.x, e.clientY - rect.y];
}

/**
 * Координаты центра элемента
 * @param el
 */
export function centerCoords(el: HTMLElement): [number, number] {
  const rect = el.getBoundingClientRect();
  return [rect.width / 2, rect.height / 2];
}

/**
 * Предупреждение в консоль с префиском `[Ripple]`
 * @param args
 */
export function consoleWarn(...args: unknown[]): void {
  // eslint-disable-next-line no-console
  console.warn('[Ripple]', ...args);
}

/**
 * Применение стилей к элементу
 */
export function applyStyles(el: HTMLElement, data: Partial<CSSStyleDeclaration>): void {
  const entries = Object.entries(data) as [
    keyof CSSStyleDeclaration,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    any
  ];

  entries.forEach(([key, value]) => {
    el.style[key] = value;
  });
}
