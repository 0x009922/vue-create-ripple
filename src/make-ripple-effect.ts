import { NormalizedRippleOptions } from './ripple-options';
import { computeRippleRadius, nextFrame, applyStyles } from './tools';

export interface Ripple {
  hide: {
    (): void;
  };
}

export interface MakeRippleEffect {
  (
    el: HTMLElement,
    options: NormalizedRippleOptions,
    x: number,
    y: number,
  ): Ripple;
}

function createContainer(opts: NormalizedRippleOptions): HTMLDivElement {
  const el = document.createElement('div');

  el.style.pointerEvents = 'none';
  el.style.position = 'absolute';
  el.style.overflow = 'hidden';
  el.style.borderRadius = 'inherit';
  el.style.transition = `opacity ${opts.disappearDuration} ${opts.disappearEasing}`;

  // Может быть так, что на контейнер распространятся правила margin/padding.
  // Поэтому отключаю экплицитно
  el.style.margin = '0';
  el.style.padding = '0';

  (
    [
      'left',
      'top',
      'right',
      'bottom'
    ] as (keyof Pick<CSSStyleDeclaration, 'left' | 'top' | 'right' | 'bottom'>)[]
  ).forEach((prop) => {
    el.style[prop] = '0';
  });

  applyStyles(el, opts.containerStyle || {});

  return el;
}

function createCircle(opts: NormalizedRippleOptions, radius: number): HTMLDivElement {
  const el = document.createElement('div');

  el.className = opts.class || '';

  el.style.borderRadius = '50%';
  el.style.transition = `transform ${opts.appearDuration} ${opts.appearEasing}`;
  el.style.transformOrigin = 'center center';
  el.style.transform = 'scale(0)';
  el.style.position = 'absolute';
  // eslint-disable-next-line no-multi-assign
  el.style.width = el.style.height = `${radius * 2}px`;

  applyStyles(el, opts.circleStyle || {});

  return el;
}

function setCirclePosition(el: HTMLElement, cx: number, cy: number, radius: number) {
  [el.style.left, el.style.top] = [cx - radius, cy - radius].map((x) => `${x}px`);
}

export const makeRippleEffect: MakeRippleEffect = (el, opts, x, y) => {
  const elRect = el.getBoundingClientRect();
  const radius = opts.radius || computeRippleRadius(elRect.width, elRect.height, x, y);
  const circle = createCircle(opts, radius);

  setCirclePosition(circle, x, y, radius);

  const container = createContainer(opts);
  container.appendChild(circle);
  el.appendChild(container);

  nextFrame(() => {
    circle.style.transform = 'scale(1)';
  });

  return {
    hide() {
      container.style.opacity = '0';
      container.addEventListener('transitionend', (ev) => {
        if (ev.target === container) {
          el.removeChild(container);
        }
      });
    }
  };
};
