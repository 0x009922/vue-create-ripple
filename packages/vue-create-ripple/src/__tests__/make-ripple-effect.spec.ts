import { mergeDefaults } from '../ripple-options';
import { makeRippleEffect } from '../make-ripple-effect';
import { computeRippleRadius } from '../tools';

// jest.mock('../tools');
// jest.mock('../tools', () => ({
//   computeRippleRadius: jest.fn(),
//   applyStyles
// }))

function createElem(): HTMLElement {
  return document.createElement('p');
}

function getContainer(root: HTMLElement): HTMLDivElement {
  return root.children[0] as HTMLDivElement;
}

function getCircle(root: HTMLElement): HTMLDivElement {
  return getContainer(root).children[0] as HTMLDivElement;
}

describe('Make ripple effect', () => {
  describe('Creation', () => {
    it('Passing default opts to container and circle', () => {
      const el = createElem();
      makeRippleEffect(el, mergeDefaults({}), 0, 0);
      expect(getComputedStyle(el.children[0] as HTMLDivElement)).toMatchSnapshot();
    });

    it('Passing styles to container', () => {
      const el = createElem();
      makeRippleEffect(el, mergeDefaults({
        containerStyle: {
          alignContent: 'center',
          transitionProperty: 'background',
          color: 'red'
        }
      }), 0, 0);
      expect(getComputedStyle(el.children[0] as HTMLDivElement)).toMatchSnapshot();
    });

    it('Passing default opts to circle', () => {
      const el = createElem();
      makeRippleEffect(el, mergeDefaults({}), 0, 0);
      const circle = el.children[0].children[0] as HTMLDivElement;
      expect(getComputedStyle(circle)).toMatchSnapshot();
    });

    it('Passing styles to circle', () => {
      const el = createElem();
      makeRippleEffect(el, mergeDefaults({
        circleStyle: {
          border: '1px solid black',
          transition: 'all .4s linear',
          borderRadius: '5px',
          opacity: '0'
        }
      }), 0, 0);

      const circle = el.children[0].children[0] as HTMLDivElement;
      expect(getComputedStyle(circle)).toMatchSnapshot();
    });

    it('Using computeRippleRadius() when radius is null', () => {
      const el = createElem();
      const RECT = {
        width: 500,
        height: 234
      } as DOMRect;
      const [X, Y] = [123, 500];
      jest.spyOn(el, 'getBoundingClientRect').mockImplementation(() => RECT);
      const radius = computeRippleRadius(RECT.width, RECT.height, X, Y);

      makeRippleEffect(el, mergeDefaults(), X, Y);

      // expect(computeRippleRadius as jest.Mock).lastCalledWith(RECT.width, RECT.height, X, Y);
      const circle = getCircle(el);
      const val = `${radius * 2}px`;
      const { width, height } = circle.style;
      expect(width).toEqual(val);
      expect(height).toEqual(val);
    });

    it('Set radius when it is not null', () => {
      const el = createElem();
      const RADIUS = 4123;

      makeRippleEffect(el, mergeDefaults({
        radius: RADIUS
      }), 0, 0);

      const circle = el.children[0].children[0] as HTMLDivElement;
      const val = `${RADIUS * 2}px`;
      const { width, height } = circle.style;
      expect(val).toEqual(width);
      expect(val).toEqual(height);
    });

    // it('Setting transform in nextFrame()', () => {
    //   const el = createElem();
    //   let callback: (() => void) | null = null;
    //   (nextFrame as jest.Mock).mockImplementation((cb) => {
    //     callback = cb;
    //     return new Promise(() => 0);
    //   });

    //   makeRippleEffect(el, mergeDefaults(), 0, 0);

    //   const circle = getCircle(el);
    //   expect(circle.style.transform).toEqual('scale(0)');
    //   expect(callback).not.toBeNull();
    //   callback!();
    //   expect(circle.style.transform).toEqual('scale(1)');
    // });
  });

  describe('Hiding', () => {
    // it('Setting container opacity to 0');

    // it('Not removing container before its transitionend');

    // it('Not removing container after circle transitionend');

    // it('Removing container after its own transitionend');
  });
});
