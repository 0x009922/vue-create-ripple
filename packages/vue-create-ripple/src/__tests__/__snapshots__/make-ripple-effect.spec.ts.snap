// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make ripple effect Creation Passing default opts to circle 1`] = `
CSSStyleDeclaration {
  "0": "display",
  "1": "visibility",
  "10": "top",
  "2": "border-radius",
  "3": "transition",
  "4": "transform-origin",
  "5": "transform",
  "6": "position",
  "7": "height",
  "8": "width",
  "9": "left",
  "_importants": Object {
    "border-radius": "",
    "display": "",
    "height": "",
    "left": "",
    "position": "",
    "top": "",
    "transform": "",
    "transform-origin": "",
    "transition": "",
    "visibility": undefined,
    "width": "",
  },
  "_length": 11,
  "_onChange": [Function],
  "_values": Object {
    "border-radius": "99999px",
    "display": "block",
    "height": "0px",
    "left": "0px",
    "position": "absolute",
    "top": "0px",
    "transform": "scale(0)",
    "transform-origin": "center center",
    "transition": "transform .8s cubic-bezier(0.16, 1, 0.3, 1)",
    "visibility": "visible",
    "width": "0px",
  },
}
`;

exports[`Make ripple effect Creation Passing default opts to container and circle 1`] = `
CSSStyleDeclaration {
  "0": "display",
  "1": "visibility",
  "10": "top",
  "11": "right",
  "12": "bottom",
  "2": "pointer-events",
  "3": "position",
  "4": "overflow",
  "5": "border-radius",
  "6": "transition",
  "7": "margin",
  "8": "padding",
  "9": "left",
  "_importants": Object {
    "border-radius": "",
    "bottom": "",
    "display": "",
    "left": "",
    "margin": "",
    "overflow": "",
    "padding": "",
    "pointer-events": "",
    "position": "",
    "right": "",
    "top": "",
    "transition": "",
    "visibility": undefined,
  },
  "_length": 13,
  "_onChange": [Function],
  "_values": Object {
    "border-radius": "inherit",
    "bottom": "0px",
    "display": "block",
    "left": "0px",
    "margin": "0px",
    "margin-bottom": "0px",
    "margin-left": "0px",
    "margin-right": "0px",
    "margin-top": "0px",
    "overflow": "hidden",
    "padding": "0px",
    "padding-bottom": "0px",
    "padding-left": "0px",
    "padding-right": "0px",
    "padding-top": "0px",
    "pointer-events": "none",
    "position": "absolute",
    "right": "0px",
    "top": "0px",
    "transition": "opacity .7s cubic-bezier(0.4, 0, 0.2, 1)",
    "visibility": "visible",
  },
}
`;

exports[`Make ripple effect Creation Passing styles to circle 1`] = `
CSSStyleDeclaration {
  "0": "display",
  "1": "visibility",
  "10": "opacity",
  "11": "left",
  "12": "top",
  "2": "border-radius",
  "3": "transition",
  "4": "transform-origin",
  "5": "transform",
  "6": "position",
  "7": "height",
  "8": "width",
  "9": "border",
  "_importants": Object {
    "border": "",
    "border-radius": "",
    "display": "",
    "height": "",
    "left": "",
    "opacity": "",
    "position": "",
    "top": "",
    "transform": "",
    "transform-origin": "",
    "transition": "",
    "visibility": undefined,
    "width": "",
  },
  "_length": 13,
  "_onChange": [Function],
  "_values": Object {
    "border": "1px solid black",
    "border-bottom": "1px solid black",
    "border-bottom-color": "black",
    "border-bottom-style": "solid",
    "border-bottom-width": "1px",
    "border-color": "black",
    "border-left": "1px solid black",
    "border-left-color": "black",
    "border-left-style": "solid",
    "border-left-width": "1px",
    "border-radius": "5px",
    "border-right": "1px solid black",
    "border-right-color": "black",
    "border-right-style": "solid",
    "border-right-width": "1px",
    "border-style": "solid",
    "border-top": "1px solid black",
    "border-top-color": "black",
    "border-top-style": "solid",
    "border-top-width": "1px",
    "border-width": "1px",
    "display": "block",
    "height": "0px",
    "left": "0px",
    "opacity": "0",
    "position": "absolute",
    "top": "0px",
    "transform": "scale(0)",
    "transform-origin": "center center",
    "transition": "all .4s linear",
    "visibility": "visible",
    "width": "0px",
  },
}
`;

exports[`Make ripple effect Creation Passing styles to container 1`] = `
CSSStyleDeclaration {
  "0": "display",
  "1": "visibility",
  "10": "top",
  "11": "right",
  "12": "bottom",
  "13": "align-content",
  "14": "transition-property",
  "15": "color",
  "2": "pointer-events",
  "3": "position",
  "4": "overflow",
  "5": "border-radius",
  "6": "transition",
  "7": "margin",
  "8": "padding",
  "9": "left",
  "_importants": Object {
    "align-content": "",
    "border-radius": "",
    "bottom": "",
    "color": "",
    "display": "",
    "left": "",
    "margin": "",
    "overflow": "",
    "padding": "",
    "pointer-events": "",
    "position": "",
    "right": "",
    "top": "",
    "transition": "",
    "transition-property": "",
    "visibility": undefined,
  },
  "_length": 16,
  "_onChange": [Function],
  "_values": Object {
    "align-content": "center",
    "border-radius": "inherit",
    "bottom": "0px",
    "color": "red",
    "display": "block",
    "left": "0px",
    "margin": "0px",
    "margin-bottom": "0px",
    "margin-left": "0px",
    "margin-right": "0px",
    "margin-top": "0px",
    "overflow": "hidden",
    "padding": "0px",
    "padding-bottom": "0px",
    "padding-left": "0px",
    "padding-right": "0px",
    "padding-top": "0px",
    "pointer-events": "none",
    "position": "absolute",
    "right": "0px",
    "top": "0px",
    "transition": "opacity .7s cubic-bezier(0.4, 0, 0.2, 1)",
    "transition-property": "background",
    "visibility": "visible",
  },
}
`;
